packages/patterns/src/signal.ts

```diff
@@ -31,41 +32,61 @@ export type Listener<T> = (data: T) => void;
  * Notice that the Signals are public.
  * We don't need to implement specific subscriptions on the class, unless we need to expose it as a remote service.
  */
-export class Signal<T> extends Set<Listener<T>> {
-    private onceHandlers = new Set<Listener<T>>();
+export class Signal<T> {
+    private handlers = new Map<Listener<T>, IS_ONCE>();
+    constructor(handlers?: Listener<T>[], once?: Listener<T>[]) {
+        handlers?.forEach((handler) => this.subscribe(handler));
+        once?.forEach((handler) => this.once(handler));
+    }
+
     /**
      * Subscribe a notification callback
      * @param handler - Will be executed with a data arg when a notification occurs
      */
     subscribe = (handler: Listener<T>) => {
-        this.add(handler);
+        this.handlers.set(handler, false);
     };
+
+    /**
+     * Subscribe to only the next notification
+     * @param handler - Will be executed with a data arg when a notification occurs
+     */
     once = (handler: Listener<T>) => {
-        this.onceHandlers.add(handler);
+        this.unsubscribe(handler);
+        this.handlers.set(handler, true);
     };
 
+    /**
+     * @returns true if a listener is subscribed
+     */
+    has(value: Listener<T>): boolean {
+        return this.handlers.has(value);
+    }
+
     /**
      * Unsubscribe an existing callback
      */
     unsubscribe = (handler: Listener<T>) => {
-        this.onceHandlers.delete(handler);
-        this.delete(handler);
+        this.handlers.delete(handler);
     };
+
+    get size(): number {
+        return this.handlers.size;
+    }
+
     /**
      * Notify all subscribers with arg data
      */
     notify = (data: T) => {
-        for (const handler of this) {
-            handler(data);
-        }
-        for (const handler of this.onceHandlers) {
+        for (const [handler, isOnce] of this.handlers) {
             handler(data);
```

Comment:
The one time handlers are new the Signal so interleaving them according to the registration order is fine, but since the event emitter now uses the Signal as base and is wildly used by other systems, this might change the order that events are handled and may lead to issues.

Reply:
I think this is a worthwhile change with low risk, don't you?

Reply:
worthy yes, low-risk no